video/cursor.h

```diff
@@ -91,63 +189,100 @@ void cursorDown() {
 	//
 	// Check if scroll required
 	//
-	if (activeCursor->Y + fontH - 1 > activeViewport->Y2) {
-		activeCursor->Y -= fontH;
-		if (~cursorBehaviour & 0x01) {
-			scrollRegion(activeViewport, 3, fontH);
-		}
-		else {
-			activeCursor->X = activeViewport->X2 + 1;
-		}
-	}
+	cursorScrollOrWrap();
 }
 
 // Move the active cursor up a line
 //
 void cursorUp() {
-	activeCursor->Y -= fontH;
-	if (activeCursor->Y < activeViewport->Y1) {
-		activeCursor->Y = activeViewport->Y1;
+	if (cursorBehaviour.flipXY) {
+		activeCursor->X += (cursorBehaviour.invertHorizontal ? fontW : -fontW);
+	} else {
+		activeCursor->Y += (cursorBehaviour.invertVertical ? fontH : -fontH);
 	}
+	cursorScrollOrWrap();
 }
 
 // Move the active cursor back one character
 //
 void cursorLeft() {
-	activeCursor->X -= fontW;											
-	if (activeCursor->X < activeViewport->X1) {						// If moved past left edge of active viewport then
-		activeCursor->X = activeViewport->X1;						// Lock it there
+	if (cursorBehaviour.flipXY) {
+		activeCursor->Y += (cursorBehaviour.invertVertical ? fontH : -fontH);
+	} else {
+		activeCursor->X += (cursorBehaviour.invertHorizontal ? fontW : -fontW);
+	}
+	if (cursorScrollOrWrap()) {
+		// wrapped, so move cursor up a line
+		cursorUp();
+	}
+}
+
+void cursorAutoNewline() {
+	if (cursorIsOffRight() && (textCursorActive() || !cursorBehaviour.grNoSpecialActions)) {
+		cursorCR();
+		cursorDown();
 	}
 }
 
 // Advance the active cursor right one character
 //
-void cursorRight() {
-	activeCursor->X += fontW;											
-	if (activeCursor->X > activeViewport->X2) {						// If advanced past right edge of active viewport
-		if (textCursorActive() || (~cursorBehaviour & 0x40)) {		// If it is a text cursor or VDU 5 CR/LF is enabled then
-			cursorCR();												// Do carriage return
-			cursorDown();											// and line feed
-		}
+void cursorRight(bool scrollProtect = false) {
```

Comment:
this function is what has to understand the `scrollProtect` mechanism.

this function will only ever be called with a `true` argument from inside the `plotCharacter` routine if scroll protect has been enabled.  all other calls to this function, such as from the cursor right VDU call (VDU 9), will use a `false` argument ensuring that scrolling/wrapping behaviour works as appropriate